### DISCUSSION TITLE
insert not working

### DISCUSSION THREAD

**Discussion by suraj-ora-2020** [sqlalchemy/sqlalchemy#10016] [Usage Questions]
**Number 1**

from sqlalchemy import create_engine,text,select,insert,update
from sqlalchemy.pool import NullPool
import oracledb
import time
import threading
import os

def creator():
    return pool.acquire(cclass="NewApp2",purity=oracledb.PURITY_SELF)

pool = oracledb.create_pool(user="sks1", password="tiger", dsn="localhost:51521/orclpdb1:pooled",
                            min=1, max=1, increment=1)

engine = create_engine("oracle+oracledb://", creator=creator, poolclass=NullPool)

conn = engine.connect()
stmt = (insert(Persons).values(personid=2,lastname='singh',firstname='jitender',address='india'))
result = conn.execute(stmt)
conn.close()

pool.close()

**Number 2**

from sqlalchemy import create_engine,text
from sqlalchemy.pool import NullPool
import oracledb
import time
import threading
import os

def creator():
    return pool.acquire(cclass="NewApp2",purity=oracledb.PURITY_SELF)

pool = oracledb.create_pool(user="sks1", password="tiger", dsn="localhost:51521/orclpdb1:pooled",
                            min=1, max=1, increment=1)

engine = create_engine("oracle+oracledb://", creator=creator, poolclass=NullPool)

conn = engine.connect()
result = conn.execute(text("insert into Persons values (2,'singh','jitender','india')"))
conn.close()

pool .close()

Both the above programs are doing the same things but first program is giving error while the second program runs fine.
**Error**
Traceback (most recent call last):
  File "/Users/apple/bofa/test_error.py", line 40, in <module>
    stmt = (insert(Persons).values(personid=2,lastname='singh',firstname='jitender',address='india'))
            ^^^^^^
NameError: name 'insert' is not defined

**Comment by CaselIT** [comment]
Hi,

Can you please correctly format the message? it's very hard to read as provided. See https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax for how to use github formatting

**Comment by suraj-ora-2020** [comment]
**Program 1**

```
from sqlalchemy import create_engine,text,select,insert,update
from sqlalchemy.pool import NullPool
import oracledb
import time
import threading
import os

def creator():
     return pool.acquire(cclass="NewApp2",purity=oracledb.PURITY_SELF)

pool = oracledb.create_pool(user="sks1", password="tiger", dsn="localhost:51521/orclpdb1:pooled",
              min=1, max=1, increment=1)

engine = create_engine("oracle+oracledb://", creator=creator, poolclass=NullPool)

conn = engine.connect()
stmt = (insert(Persons).values(personid=2,lastname='singh',firstname='jitender',address='india'))
result = conn.execute(stmt)
conn.close()

pool.close()
```

**Program 2**

```
from sqlalchemy import create_engine,text
from sqlalchemy.pool import NullPool
import oracledb
import time
import threading
import os

def creator():
     return pool.acquire(cclass="NewApp2",purity=oracledb.PURITY_SELF)

pool = oracledb.create_pool(user="sks1", password="tiger", dsn="localhost:51521/orclpdb1:pooled",
            min=1, max=1, increment=1)

engine = create_engine("oracle+oracledb://", creator=creator, poolclass=NullPool)

conn = engine.connect()
result = conn.execute(text("insert into Persons values (2,'singh','jitender','india')"))
conn.close()

pool .close()

```

**Both the above programs are doing the same things but first program is giving error while the second program runs fine.**

**Error**
```
Traceback (most recent call last):
File "/Users/apple/bofa/test_error.py", line 40, in
stmt = (insert(Persons).values(personid=2,lastname='singh',firstname='jitender',address='india'))
^^^^^^
NameError: name 'insert' is not defined
```

**Reply by CaselIT** [reply]
As provided both programs should work. That error means that insert was not imported. Ensure that it is in the file where you use it

**Comment by suraj-ora-2020** [comment]
In program 1 the first line only imports all those functions
from sqlalchemy import create_engine,text,select,insert,update

**Reply by CaselIT** [reply]
yes, but program 1 works if you define a table called `Persons`. Verify that in your real program you are importing it or you aren't assigning insert to a variable.

A common practice is to use `import sqlalchemy as sa` so you can then use `sa.insert(...)`
